### PR TITLE
SW-3908 Warn When Undoing Planting Complete will Delete Stats

### PR DESCRIPTION
**Pull request by charlie-tf**
- determine the zone for a subzone when clicking map
- select whether that zone has statistics
- if there are no statistics then allow update without a warning
- if there are statistics:
  - open warning modal
  - set a callback that is executed if the deletion is confirmed
- after update refresh observation results as well as planting site

image

### CODE REVIEW

File: src/components/NurseryWithdrawals/PlantingProgressMap.tsx [hunk 1]
```diff
@@ -73,48 +86,74 @@ export default function PlantingProgressMap({ plantingSiteId }: PlantingProgress
       if (updateStatus.status === 'error') {
         snackbar.toastError(strings.GENERIC_ERROR);
       } else if (updateStatus.status === 'success') {
-        // refresh planting site data to get new completed state for subzone
+        // refresh planting site and observations data
         dispatch(requestPlantingSites(org.selectedOrganization.id));
+        dispatch(requestObservationsResults(org.selectedOrganization.id));
```
Comment (constanzauanini) [inline src/components/NurseryWithdrawals/PlantingProgressMap.tsx:91]: There is a function _reloadTrackingAndObservations_ in the parent (Planting progress tab content) that refreshes this. Maybe we can pass it to the map as we are passing it to the list? It also refreshes plantingSites

Comment (charlie-tf) [reply src/components/NurseryWithdrawals/PlantingProgressMap.tsx:91]: Oh, I'll take a look at using that 👍

Comment (charlie-tf) [reply src/components/NurseryWithdrawals/PlantingProgressMap.tsx:91]: I'm not sure why but it caused the map to glitch like crazy. I'm going to leave as is for now, and we can possibly follow up.

**Review by constanzauanini — APPROVED** [review]
Looks good otherwise !

File: src/components/NurseryWithdrawals/PlantingProgressMap.tsx (hunk 1 above)
Comment (karthikbtf) [reply src/components/NurseryWithdrawals/PlantingProgressMap.tsx:91]: I will take a look at this today.

File: src/components/NurseryWithdrawals/PlantingProgressMap.tsx [hunk 2]
```diff
@@ -92,29 +106,63 @@ export default function PlantingProgressMap({ plantingSiteId }: PlantingProgress
...
+    },
+    [selectedZoneHasStats, completeUpdate]
+  );
+
+  const onOpenMapDialog = useCallback(
+    (id: number) => {
+      const selectedZone = plantingSite?.plantingZones?.find((zone) =>
+        zone.plantingSubzones.map((sz) => sz.id).includes(id)
+      );
+      setZoneIdSelected(selectedZone?.id ?? -1);
```
Comment (karthikbtf) [inline src/components/NurseryWithdrawals/PlantingProgressMap.tsx:132]: do we need the -1 fallback?

File: src/components/NurseryWithdrawals/PlantingProgressMap.tsx [hunk 3]
```diff
@@ -73,15 +86,16 @@ export default function PlantingProgressMap({ plantingSiteId }: PlantingProgress
...
         dispatch(requestPlantingSites(org.selectedOrganization.id));
+        dispatch(requestObservationsResults(org.selectedOrganization.id));
       }
     }
   }, [updateStatus, dispatch, snackbar, org.selectedOrganization.id]);
 
-  const updatePlantingComplete = useCallback(
-    (id: number, val: boolean) => {
-      // TODO: warn if undoing planting complete will erase statistics
+  const [confirmDeleteCb, setConfirmDeleteCb] = useState<() => void>(() => () => null);
```
Comment (karthikbtf) [inline src/components/NurseryWithdrawals/PlantingProgressMap.tsx:96]: can we leverage onSubmit to call updatePlantingComplete, without this additional callback?

File: src/components/NurseryWithdrawals/PlantingProgressMap.tsx [hunk 4]
```diff
@@ -92,29 +106,63 @@ export default function PlantingProgressMap({ plantingSiteId }: PlantingProgress
...
     [dispatch]
   );
 
+  const updatePlantingComplete = useCallback(
+    (id: number, val: boolean) => {
+      if (!selectedZoneHasStats) {
+        completeUpdate(id, val)();
+      } else {
+        setConfirmDeleteCb(() => completeUpdate(id, val));
+        setStatsWarningDialogOpen(true);
```
Comment (karthikbtf) [inline src/components/NurseryWithdrawals/PlantingProgressMap.tsx:121]: suggestion:
remove the callback approach, find the zone id here and set that as a property.
if the zone id is valid then show the modal (pass the zone id to the modal)
modal can send the zone id back as part of the onsubmit.
we can remove a bunch of additional state/message-passing

File: src/components/NurseryWithdrawals/PlantingProgressMapDialog.tsx [hunk 5]
```diff
@@ -44,6 +46,12 @@ export default function PlantingProgressMapDialog({
     return Object.values(species).reduce((prev, curr) => prev + curr, 0);
   }, [species]);
 
+  useEffect(() => {
```
Comment (karthikbtf) [inline src/components/NurseryWithdrawals/PlantingProgressMapDialog.tsx:49]: i think we can remove this.. we can find the zone id at the time of deciding whether to open the modal.. and pass that to the modal.. which can pass it back onSubmit

**Review by karthikbtf — COMMENTED** [review]
Might be better to remove the callback and the onOpen.. find the zone id when we decide to show the modal and set that as a property. Modal can pass it back in onSubmit

File: src/components/NurseryWithdrawals/PlantingProgressMap.tsx (hunk 2 above)
Comment (constanzauanini) [reply src/components/NurseryWithdrawals/PlantingProgressMap.tsx:132]: It could be undefined instead of -1 maybe

**Comment by karthikbtf** [comment]
Merging this in for now.
Charlie when you get a chance review https://github.com/terraware/terraware-web/pull/1601 which has my suggestions implemented.

**Comment by karthikbtf** [comment]
i merged the wrong PR :(

**Pull request merged**